domain/model/service/service.go

```diff
@@ -709,3 +729,56 @@ func (s WatchableService) WatchModel(ctx context.Context, modelUUID coremodel.UU
 		}),
 	)
 }
+
+// WatchModelCloudCredential returns a new NotifyWatcher watching for changes that
+// result in the cloud spec for a model changing. The changes watched for are:
+// - updates to model cloud.
+// - updates to model credential.
+// - changes to the credential set on a model.
+// The following errors can be expected:
+// - [modelerrors.NotFound] when the model is not found.
+func (s *WatchableService) WatchModelCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (watcher.NotifyWatcher, error) {
+	if err := modelUUID.Validate(); err != nil {
+		return nil, errors.Errorf("invalid model UUID watching model cloud credential: %w", err)
+	}
+
+	// Get the model's cloud and credential UUID to use in the watcher filters.
+	cloudUUID, credentialUUID, err := s.st.GetModelCloudAndCredential(ctx, modelUUID)
```

Comment:
credentialUUID needs to be protected it looks like. At the moment we have unprotected read and writes.

Reply:
I had thought the reads and writes were sequential due to how the watcher operates but a closer look shows that's probably not the case, so added atomic access.